extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/SmallRyeFaultToleranceProcessor.java

```diff
@@ -187,13 +192,42 @@ public void transform(TransformationContext ctx) {
     @BuildStep
     // needs to be RUNTIME_INIT because we need to read MP Config
     @Record(ExecutionTime.RUNTIME_INIT)
-    void validateFaultToleranceAnnotations(
-            ValidationPhaseBuildItem validationPhase, SmallryeFaultToleranceRecorder recorder) {
-        List<String> beanNames = new ArrayList<>();
-        for (BeanInfo bean : validationPhase.getContext().beans().classBeans()) {
-            beanNames.add(bean.getBeanClass().toString());
+    void validateFaultToleranceAnnotations(SmallryeFaultToleranceRecorder recorder,
+            ValidationPhaseBuildItem validationPhase) {
+        AnnotationStore annotationStore = validationPhase.getContext().get(BuildExtension.Key.ANNOTATION_STORE);
+        Set<String> beanNames = new HashSet<>();
+        IndexView index = combinedIndexBuildItem.getIndex();
+
+        for (BeanInfo info : validationPhase.getContext().beans()) {
+            if (hasFTAnnotations(annotationStore, info.getImplClazz())) {
+                beanNames.add(info.getBeanClass().toString());
+            }
         }
-        recorder.validate(beanNames);
+
+        recorder.createFaultToleranceOperation(beanNames);
+    }
+
+    private boolean hasFTAnnotations(AnnotationStore annotationStore, ClassInfo info) {
+        // first check annotations on type
+        if (annotationStore.hasAnyAnnotation(info, FT_ANNOTATIONS)) {
+            return true;
+        }
+
+        // then check on the methods
+        for (MethodInfo method : info.methods()) {
+            if (annotationStore.hasAnyAnnotation(method, FT_ANNOTATIONS)) {
+                return true;
+            }
+        }
+
+        // then check on the parent
+        DotName parentClassName = info.superName();
+        ClassInfo parentClassInfo = combinedIndexBuildItem.getIndex().getClassByName(parentClassName);
```

Comment:
I think that using `BeanArchiveIndexBuildItem` would be more appropriate in this case. Also the injected fields of `SmallRyeFaultToleranceProcessor` should be probably turned into injected params of the `SmallRyeFaultToleranceProcessor.build()` build step (they're not shared so it does not make much sense to use fields). I know that it's not a subject of this PR but it would deserve a small refactoring ;-).

Reply:
Switch to `BeanArchiveIndexBuildItem`.
Refactor to inject everything via parameters.